### PR TITLE
Add Slab MCP with ability to have internal MCP servers with bearer to…

### PR DESCRIPTION
**Pull request by frankaloia**
## Description

* Added topic and post read ability for Slab using their GraphQL endpoint
* Slab knowledge base requires API tokens to make API calls (https://help.slab.com/en/articles/6545629-developer-tools-api-webhooks)
* Added ability to save bearer token secret for MCP servers in front. Created new table to store information. Minor UX + API changes to facilitate this. @Fraggle I know this has been a discussion in the past, so open to alternative paths if there was already a different identified solution.
* Open to splitting this into 2 PRs, but didn't see the value since we wouldn't be able to test without 1 MCP server anyway (and it's behind FF)

## Tests

1. Tested all tools and edge cases
2. See screenshots for UX changes

Screenshot 2025-11-19 at 2 09 27 PM
Screenshot 2025-11-19 at 2 09 34 PM

## Risk

* None intended, but does change existing MCP logic

## Deploy Plan

1. Run migration
2. Deploy front

### CODE REVIEW

File: front/lib/actions/mcp_internal_actions/servers/slab/slab_response_helpers.ts [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+  topicMap.forEach((topic) => {
+    if (topic.parentId) {
+      const parent = topicMap.get(topic.parentId);
+      if (parent) {
+        const updatedParent = {
+          ...parent,
+          children: [...(parent.children ?? []), topic],
+        };
+        topicMap.set(topic.parentId, updatedParent);
```
[the remark this answers is not in the document]
Comment (frankaloia) [reply front/lib/actions/mcp_internal_actions/servers/slab/slab_response_helpers.ts]: Agreed

File: front/components/actions/mcp/CreateMCPServerSheet.tsx [hunk 2]
```diff
@@ -248,6 +251,38 @@ export function CreateMCPServerSheet({
         return;
       }
       server = createRes.value.server;
+
+      // Save manual auth credentials if this server supports manual auth
```
Comment (Fraggle) [inline front/components/actions/mcp/CreateMCPServerSheet.tsx]: Is it really about manual auth ?

File: front/lib/resources/internal_mcp_server_credentials_resource.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import { Err, Ok, redactString } from "@app/types";
+
+const SECRET_REDACTION_COOLDOWN_IN_MINUTES = 10;
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface, @typescript-eslint/no-unsafe-declaration-merging
+export interface InternalMCPServerCredentialResource
+  extends ReadonlyAttributesType<InternalMCPServerCredentialModel> {}
+
+// eslint-disable-next-line @typescript-eslint/no-unsafe-declaration-merging
+export class InternalMCPServerCredentialResource extends BaseResource<InternalMCPServerCredentialModel> {
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_credentials_resource.ts]: You don't need to create a resource for "sub-entity" that can only live within another one.
You can put the needed methods directly in InternalMCPServerInMemoryResource

File: front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts [hunk 4]
```diff
@@ -138,9 +164,30 @@ async function handler(
...
+            mcpServerView.mcpServerId
+          );
+        const credentialJSON = credentials?.toJSON();
+        if (credentialJSON) {
+          serverView = {
+            ...viewJSON,
+            server: {
+              ...viewJSON.server,
+              sharedSecret: credentialJSON.sharedSecret ?? null,
+              customHeaders: credentialJSON.customHeaders ?? null,
```
Comment (Fraggle) [inline front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: So if I understand correctly, only this endpoint will return the sharedSecret & customHeaders, the `toJSON` of `InternalMCPServerInMemoryResource` will return null for theses properties even if they are actually set for an internal mcp server ?

Comment (Fraggle) [reply front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: I'm not even sure that the currrent implementation will pick up the 2 values when actually calling the mcp server in the agentic loop, does it ?

Comment (Fraggle) [reply front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: Ah just saw that you also fetch it [here](https://github.com/dust-tt/dust/pull/18096/files#diff-8528ddf58baaa2dd35b473c43d05b017ae1199a379e53133ca0b77836f38574fR188)

Comment (Fraggle) [reply front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: I'd rather have it done once and for all in `getCachedMetadata` of `InternalMCPServerInMemoryResource`, wdyt ?

**Review by Fraggle — APPROVED** [review]
It's a lot of changes so it's hard to be sure it really works but i trust you™

I made a suggestion regarding the fetching of the informations, let me know if it isn't clear.

File: front/lib/resources/internal_mcp_server_credentials_resource.ts (hunk 3 above)
Comment (Fraggle) [reply front/lib/resources/internal_mcp_server_credentials_resource.ts]: (i learned it recently)

Comment (Fraggle) [reply front/lib/resources/internal_mcp_server_credentials_resource.ts]: By moving the methods in there them, you'll be able to have them available for the toJSON() as well as getter on the `InternalMCPServerInMemoryResource`.

File: front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts (hunk 4 above)
Comment (Fraggle) [reply front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: Actually, getCachedMetadata is about the general server informations, not the informations specific to a particular instance.
But my point is to load it somehow within the `InternalMCPServerInMemoryResource` instance so it's always available without fetching it manually in specific places.

Comment (Fraggle) [reply front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: Thinking more about, might be okay to get it on demand and cache it on instance since it's a very rare occurrence.

```ts
let _sharedSecret :string|undefined|null;

get sharedSecret():string|null {
  if (_sharedSecret === undefined) {
    // Fetch and store on _sharedSecret with a potential null value.
  }
  return _sharedSecret;
}

toJSON() {
  ...,
  sharedSecret: this.sharedSecret
}
```

File: front/lib/resources/internal_mcp_server_credentials_resource.ts (hunk 3 above)
Comment (frankaloia) [reply front/lib/resources/internal_mcp_server_credentials_resource.ts]: @Fraggle I have some concern with putting it in InternalMCPServerInMemoryResource given that it then bloats the purpose of that data type. Currently, that data type is really centered around managing our static MCP definition per workspace. By adding this, it would now store workspace-specific auth information.

If you are not viewing that as a concern, I'm open to it.

File: front/components/actions/mcp/CreateMCPServerSheet.tsx (hunk 2 above)
Comment (frankaloia) [reply front/components/actions/mcp/CreateMCPServerSheet.tsx]: No, can reword

File: front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts (hunk 4 above)
Comment (frankaloia) [reply front/pages/api/w/[wId]/spaces/[spaceId]/mcp_views/index.ts]: Added a comment related to InternalMCPServerInMemoryResource above

File: front/lib/resources/internal_mcp_server_credentials_resource.ts (hunk 3 above)
Comment (Fraggle) [reply front/lib/resources/internal_mcp_server_credentials_resource.ts]: > Currently, that data type is really centered around managing our static MCP definition per workspace

I'm not 100% sure what you mean but it was designed as the counterpart of `RemoteMCPServerResource` which store workspace-specific auth stuff.
The fact that it do not hold any so far in InternalXXX is just because we didn't need it but imho, it's fine.
The id passed to create an instance is already specific to a workspace.

Comment (frankaloia) [reply front/lib/resources/internal_mcp_server_credentials_resource.ts]: Ok, i am going to take all of your suggestion then 👍

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts [hunk 5]
```diff
@@ -352,11 +368,135 @@ export class InternalMCPServerInMemoryResource {
     return removeNulls(resources);
   }
 
+  static async fetchInternalServerCredential(auth: Authenticator, id: string) {
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_in_memory_resource.ts]: nit: I would make private and not static.

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts [hunk 6]
```diff
@@ -352,11 +368,135 @@ export class InternalMCPServerInMemoryResource {
...
+    if (r.isErr()) {
+      return new Err(
+        new DustError(
+          "unauthorized",
+          `Invalid internal MCP server ID: ${this.id}`
+        )
+      );
+    }
+
+    const workspaceModelId = r.value.workspaceModelId;
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_in_memory_resource.ts]: Also, you can use the workspaceId from the authenticator (getNonNullableWorkspace()).

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts [hunk 7]
```diff
@@ -352,11 +368,135 @@ export class InternalMCPServerInMemoryResource {
...
+      }
+
+      if (Object.keys(updatePayload).length > 0) {
+        await existing.update(updatePayload);
+      }
+
+      record = existing;
+    } else {
+      record = await InternalMCPServerCredentialModel.create({
+        workspaceId: workspaceModelId,
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_in_memory_resource.ts]: same here

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts [hunk 8]
```diff
@@ -352,11 +368,135 @@ export class InternalMCPServerInMemoryResource {
...
+      return new Err(
+        new DustError(
+          "unauthorized",
+          "The user is not authorized to update this MCP server."
+        )
+      );
+    }
+
+    const r = getInternalMCPServerNameAndWorkspaceId(this.id);
+    if (r.isErr()) {
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_in_memory_resource.ts]: I believe here you can assume that the server exists as you are working on an instance that has been instanciated via the ID so no need for lines 406-414.

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts [hunk 9]
```diff
@@ -352,11 +368,135 @@ export class InternalMCPServerInMemoryResource {
...
+  private getRedactedCredentials(): {
+    sharedSecret: string | null;
+    customHeaders: Record<string, string> | null;
+  } {
+    if (!doesInternalMCPServerSupportBearerToken(this.id)) {
+      return { sharedSecret: null, customHeaders: null };
+    }
+
+    if (!this._internalServerCredential) {
+      return { sharedSecret: null, customHeaders: null };
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_in_memory_resource.ts:442]: Shouldn't we throw here instead ?
Is it "support" or "require" btw ?

File: front/lib/actions/mcp_internal_actions/constants.ts [hunk 10]
```diff
@@ -1289,6 +1290,33 @@ The directive should be used to display a clickable version of the agent name in
...
   },
+  slab: {
+    id: 43,
+    availability: "manual",
+    allowMultipleInstances: true,
+    isRestricted: ({ featureFlags }) => {
+      return !featureFlags.includes("slab_mcp");
+    },
+    isPreview: true,
+    supportsBearerToken: true,
```
Comment (Fraggle) [inline front/lib/actions/mcp_internal_actions/constants.ts]: is it support or require ?
Also, it's a bit confusing because later you say "if (authorization && !bearerToken...)"

File: front/lib/actions/mcp_metadata.ts [hunk 11]
```diff
@@ -178,12 +182,45 @@ export const connectToMCPServer = async (
...
+                  connectionType: "workspace",
+                },
+              };
+
+              client.setAuthInfo(authInfo);
+              server.setAuthInfo(authInfo);
+              bearerTokenApplied = !!bearerTokenCredentials.sharedSecret;
+            }
+
+            const metadata = extractMetadataFromServerVersion(
```
Comment (Fraggle) [inline front/lib/actions/mcp_metadata.ts]: I would mimic the code we have below for remoteMCPServer ([here](https://github.com/dust-tt/dust/blob/ac599b47c3445d6271549be5988da8a71ff8cdd1/front/lib/actions/mcp_metadata.ts#L260))

- instanciate the InternalMCPServerInMemoryResource
- check if it has a bearer, use to authenticate
- if not, check authorization

so if (...) {} else if (...) {}

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts [hunk 12]
```diff
@@ -89,6 +91,8 @@ export class InternalMCPServerInMemoryResource {
     ...extractMetadataFromServerVersion(undefined),
     tools: [],
   };
+  private _internalServerCredential: InternalMCPServerCredentialModel | null =
```
Comment (Fraggle) [inline front/lib/resources/internal_mcp_server_in_memory_resource.ts]: the _ was suggested if you loaded it "on demand" in a getter, if not, you can just use without it.

**Review by Fraggle — APPROVED** [review]
Looks good :)

File: front/pages/api/w/[wId]/mcp/[serverId]/index.ts [hunk 13]
```diff
@@ -143,19 +151,30 @@ async function handler(
 
       const { serverType } = getServerTypeAndIdFromSId(serverId);
 
-      if (serverType !== "remote") {
```
Comment (Fraggle) [inline front/pages/api/w/[wId]/mcp/[serverId]/index.ts:146]: Could you update the tests to express that it should now work for internal mcp server.

File: front/lib/actions/mcp_internal_actions/constants.ts (hunk 10 above)
Comment (frankaloia) [reply front/lib/actions/mcp_internal_actions/constants.ts]: Yea in retrospect I don't like how I did this. I think it should be requires and exist inside the authorization configuration object

File: front/lib/resources/internal_mcp_server_in_memory_resource.ts (hunk 9 above)
Comment (frankaloia) [reply front/lib/resources/internal_mcp_server_in_memory_resource.ts:442]: I think I'm too inconsistent on this. It is required for this tool, but I think it will be optional for others (like HTTP tool). Will make this more clear.
